proposals/required-members.md

```diff
@@ -263,42 +263,131 @@ init_argument_initializer
 A type with a parameterless constructor that advertises a _contract_ is not allowed to be substituted for a type parameter constrained to `new()`, as there is no way
 for the generic instantiation to ensure that the requirements are satisfied.
 
+### Overriding, Hiding, and Inheriting
+
+It is an error to mark a member required if the member is not at least as visible as the containing member's type. This means the following cases are not allowed:
+
+```cs
+interface I
+{
+    int Prop1 { get; }
+}
+public class Base
+{
+    public virtual int Prop2 { get; }
+
+    protected required int _field; // Error: _field is not at least as visible as Base. Open question below about the protected constructor scenario
+    protected Base() { }
+}
+public class Derived : Base, I
+{
+    required int I.Prop1 { get; } // Error: explicit interface implementions cannot be required as they cannot be set in an object initializer
+
+    public required override int Prop2 { get; } // Error: this property is hidden by Derived.Prop2 and cannot be set in an object initializer
+    public new int Prop2 { get; }
+}
+```
+
+It is an error to hide a `required` member, as that member can no longer be set by a consumer.
+
+When overriding a `required` member, the `required` keyword must be included on the method signature. This is done so that if we ever want to allow
+unrequiring a property with an override in the future, we have design space to do so.
+
+Overrides are allowed to mark a member `required` where it was not `required` in the base type. A member so-marked is added to the required members
+list of derived type.
+
+### Metadata Representation
+
+The following 2 attributes are known to the C# compiler and required for this feature to function:
+
+```cs
+namespace System.Runtime.CompilerServices;
+
+[AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct, AllowMultiple = false)]
+public sealed class RequiredMembersAttribute : Attribute
+{
+    public RequiredMembersAttribute(string[] members) => Members = members;
+    public string[] Members { get; }
```

Comment:
It might be worthwhile to discuss in an "alternatives" section why this encoding is preferable to putting a `[Required]` attribute straight on the members which were `required` in source, and perhaps putting some attribute on the containing type to indicate "this type contains required members", similar to types that contain extension methods.